copy/copy.go

```diff
@@ -633,6 +669,26 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 	// If encrypted and decryption keys provided, we should try to decrypt
 	ic.diffIDsAreNeeded = ic.diffIDsAreNeeded || (isEncrypted(src) && ic.c.ociDecryptConfig != nil) || ic.c.ociEncryptConfig != nil
 
+	options.OptimizeDestinationImageAlreadyExists = true // TODO: PR: Remove forced enablement for PR testing
+	// If enabled, fetch and compare the destination's manifest. And as an optimization skip updating the destination iff equal
+	// To be enabled, options.OptimizeDestinationImageAlreadyExists must be true, the image must not be signed or encryted. And must
+	// not have any pending manifest updates
+	isSigned := len(sigs) > 0 || options.SignBy != "" // TODO: Consider allowing signatures updates only and skipping the image's layers/manifest copy if possible
+	isEncryptionConfigured := c.ociDecryptConfig != nil || c.ociEncryptConfig != nil
+	if options.OptimizeDestinationImageAlreadyExists && !isSigned && !isEncryptionConfigured && !ic.noPendingManifestUpdates() {
+
+		isSrcDestManifestEqual, retManifest, retManifestType, retManifestDigest, err := compareImageDestinationManifestEqual(ctx, options, src, c.dest)
+		if err != nil {
+			logrus.Debugf("Failed to compare destination image manifest: %v", err)
+			return nil, "", "", err
+		}
+
+		if isSrcDestManifestEqual {
+			c.Printf("Destination copy skipped as expected image already present\n")
+			return retManifest, retManifestType, retManifestDigest, nil
+		}
+	}
+
 	if err := ic.copyLayers(ctx); err != nil {
```

Comment:
Is it a concern that `ic.copyLayers()` updates `ic. manifestUpdates`, and so can change the result of `ic.noPendingManifestUpdates()`?  But the check to skip the copy on [L678](https://github.com/containers/image/pull/1041/files#diff-5f78baafa6d34438aba66ed64f4b16d5R678) will never see those updates?

I assume it isn't, since if no layers need to be copied because the manifest is equivalent, then `ic.copyLayers()` would not update `ic.manifestUpdates`.

But I also can't say I am a fan of this implicit contract/coupling between these bits of functionality.

Reply:
Yes, it is “a concern”.

It would of course be nicer if we could rely on the `ic.manifestUpdates` coming out of `ic.copyLayers` — but avoiding `ic.copyLayers` is the whole point, we can’t make some of them without initiating the layer reads via `GetBlob`.

Still, it would probably be worth it to make this a bit harder to miss, maybe by adding a
```go
// WARNING: If you are adding new reasons to change ic.manifestUpdates, update also the OptimizeDestinationImageAlreadyExists short-circuit conditions
```
before the update at the end of `ic.copyLayers`, and a similar
```go
// WARNING: If you are adding new reasons to change the blob, …
```
somewhere in `copyBlobFromStream`, maybe either immediately after the “The copying happens through a pipeline of connected io.Readers.” comment or close to (above?) the “Deal with layer compression/decompression if necessary” one.

Reply:
Above added